trino-aws-proxy-spi/src/main/java/io/trino/aws/proxy/spi/credentials/IdentityCredential.java

```diff
@@ -0,0 +1,42 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.aws.proxy.spi.credentials;
+
+import java.util.Optional;
+
+import static java.util.Objects.requireNonNull;
+
+public record IdentityCredential(Credential emulated, Optional<Identity> identity)
+{
+    public IdentityCredential
+    {
+        requireNonNull(emulated, "emulated is null");
+        requireNonNull(identity, "identity is null");
+    }
+
+    public IdentityCredential(Credential emulatedCredential, Identity identity)
+    {
+        this(emulatedCredential, Optional.of(identity));
+    }
+
+    public IdentityCredential(Credential emulatedCredential)
+    {
+        this(emulatedCredential, Optional.empty());
+    }
+
+    public static IdentityCredential of(Credential emulatedCredential)
```

Comment:
nit: I would decide on a single way for public clients to create instances. At the moment, there is a public constructor and a factory method with the same signature.